### PR TITLE
feat: FCM v1

### LINKED ISSUE #294: Update to new FCM API before June 2024
The current [FCM API is deprecated and will be removed in June 2024](https://firebase.google.com/docs/cloud-messaging). See this issue for details, and folks building an updated FCM Rust library we can switch to: https://github.com/panicbit/fcm-rust/issues/32

### PR DESCRIPTION
**Pull request by chris13524**
# Description

Adds support for the new FCM API using a fork of the current FCM crate: rj76/fcm-rust

Resolves #294 

Remaining work:
- [x] Add a new database column for the new credentials file in string format
- [x] Add `ECHO_TEST_FCM_V1_CREDENTIALS` secret
- [x] Validate credentials in `update_fcm_v1` route
- [ ] ~~Enumerate and handle error cases:~~ https://github.com/WalletConnect/push-server/issues/321
  - [ ] ~~Key parse as RSA~~
  - [ ] ~~Access token status code~~
  - [ ] ~~Unauthorized (?) on FCM send~~
- [x] Handle errors in fcm_v1.rs#send_notification inside `match result` by exporting necessary fields
- [ ] ~~Add length constraints for `fcm_v1_credentials`, `fcm_api_key`, `apns_topic`, `apns_certificate`, and `apns_certificate_password`, `apns_key_id`, `apns_team_id`, `apns_pkcs8_pem`~~ https://github.com/WalletConnect/push-server/issues/322
- [ ] ~~Complete full FCM payload w/ fallbacks i.e. Android, web push~~ https://github.com/WalletConnect/push-server/issues/new
- [x] When constructing the FCM provider, check if the new column is present. If so use the new FCM API
- [x] Add support to FCM library to read credentials from string: https://github.com/rj76/fcm-rust/issues/5
- [x] While a rate limit is not mentioned [in the docs](https://developers.google.com/identity/protocols/oauth2/service-account#makingrequest), we should refactor the provider code to reuse access tokens rather than issuing a new one for every push notification. (The access token appears to be valid for 1 hour.)
- [x] Reuse HTTP client for both FCM API and gauth API

~~To ship:~~ https://github.com/WalletConnect/push-server/issues/324
- [ ] ~~Work with Cloud team to add ability to configure new credentials and add necessary endpoints to Push Server~~
- [ ] ~~Consider Cloud-facing API: should `enabled_providers` include both `fcm` and `fcm_v1`? Should there be a separate endpoint `update_fcm_v1` to update? Right now yes~~
- [ ] ~~Inform users to setup new credentials~~

## How Has This Been Tested?

TBD

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

**Review by nopestack — DISMISSED** [review]

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -81,7 +81,7 @@ jobs:
         id: tf-apply
         uses: WalletConnect/actions/terraform/apply/@1.0.3
         env:
-          GRAFANA_AUTH: ${{ steps.grafana-get-key.outputs.key }}
+          TF_VAR_grafana_auth: ${{ steps.grafana-get-key.outputs.key }}
```
Comment (chris13524) [inline .github/workflows/cd.yml:84]: Avoids this error in newer TF versions and matches other repos:

```
╷
│ Error: Missing required argument
│ 
│   with provider["registry.terraform.io/grafana/grafana"],
│   on provider.tf line 16, in provider "grafana":
│   16: provider "grafana" {
│ 
│ "auth": one of `auth,cloud_api_key,oncall_access_token,sm_access_token` must be specified
╵
╷
│ Error: Missing required argument
│ 
│   with provider["registry.terraform.io/grafana/grafana"],
│   on provider.tf line 17, in provider "grafana":
│   17:   url = "https://${var.grafana_endpoint}"
│ 
│ "url": all of `auth,url` must be specified
╵
error: Recipe `tf-validate` failed with exit code 1
```

**Comment by chris13524** [comment]
@geekbrother @nopestack asking for an initial review to see if any high-level issues. There are still a few minor tasks I need to figure out as mentioned in the PR description

**Review by geekbrother — COMMENTED** [review]
While it's a draft it looks good to me! Thanks for updating to v1 👍

File: src/blob.rs [hunk 2]
```diff
@@ -16,12 +16,10 @@ pub struct DecryptedPayloadBlob {
 }
 
 impl DecryptedPayloadBlob {
```
Comment (chris13524) [inline src/blob.rs:18]: Remove unused function and improve error specificity

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -85,7 +85,7 @@ jobs:
             rustc: nightly
           - name: "Unit Tests"
             cmd: test
-            args: --features multitenant,analytics,geoblock,functional_tests
+            args: --features multitenant,analytics,geoblock,functional_tests,apns_tests,fcm_tests,fcmv1_tests
```
Comment (chris13524) [inline .github/workflows/ci.yml:88]: Made into separate features to enable running tests locally without all of these test secrets.

File: src/providers/fcm_v1.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                        priority: Some(AndroidMessagePriority::High),
+                        collapse_key: None,
+                        ttl: None,
+                        restricted_package_name: None,
+                        data: None,
+                        notification: None,
+                        fcm_options: None,
+                        direct_boot_ok: None,
+                    }),
+                    // TODO
```
Comment (chris13524) [inline src/providers/fcm_v1.rs:59]: This file missing a lot of options; will be corrected in follow-up PR

**Review by geekbrother — APPROVED** [review]

**Pull request merged**